[WC-109] Refactor progress bar web widget into pluggable widget

Keraito opened the pull request:
## What

Refactor progress bar web widget from custom React container framework into a pluggable widget.

Additional changes that are related:
- Every property related to animations has been moved towards a design property.
- Values can be static, dynamic, or expressions.

## Testing Comments
- There was a flaky e2e on click test before. But that was targeting logic from the old react container approach. So that should not be an issue anymore.
- The current test project was from the old widget so likely should be revisited entirely.

Keraito commented on packages/pluggableWidgets/progress-bar-web/src/components/__tests__/__snapshots__/ProgressBar.spec.tsx.snap:
```diff
@@ -2,14 +2,14 @@
 exports[`Progress bar has progress bar structure 1`] = `
 <div
-  className="widget-progress-bar"
+  className="widget-progress-bar progress-bar-medium"
```
@diego-antonelli one of the requirements was to not change the structure, are these css classname changes fine?

diego-antonelli replied:
I think is fine, the original widget-progress-bar is being kept :)

Keraito commented on packages/pluggableWidgets/progress-bar-web/src/components/ProgressBar.tsx:
```diff
@@ -0,0 +1,73 @@
+}: ProgressBarProps): ReactElement {
+    const errorMessage = getValuesErrorMessage(currentValue, minValue, maxValue);
+    return (
+        <div className={classNames("widget-progress-bar", "progress-bar-medium", className)} style={style}>
+            <div
+                className={classNames("progress", {
+                    "widget-progress-bar-alert": maxValue < 1,
+                    "widget-progress-bar-clickable": !!onClick
+                })}
+                onClick={onClick}
+            >
+                <div
+                    className={classNames("progress-bar", "progress-bar-default", "progress-bar-medium", className)}
```
@diego-antonelli also not quite sure whether this is a good way to implement design properties, because basically all design properties classes are now applied to both this element as well as L51. But different design properties do cover different elements, so it's kind of necessary. I sat with Jordan and we thought it was fine since we can use CSS selectors to target one and not the other, but can you check whether this is fine?

Keraito replied:
So I should remove `className` on this line (60) and then change the CSS selectors to target this element through the element on L51?

diego-antonelli replied:
Yeap

diego-antonelli commented on packages/pluggableWidgets/progress-bar-web/src/components/__tests__/ProgressBar.spec.tsx:
```diff
@@ -0,0 +1,215 @@
+        expect(progressBar).toMatchSnapshot();
+    });
+
+    it("should show positive progress", () => {
```
`shows a positive progress`

diego-antonelli commented on packages/pluggableWidgets/progress-bar-web/src/components/__tests__/ProgressBar.spec.tsx:
```diff
@@ -0,0 +1,215 @@
+        expect(getProgressBarStyle(progressBar)).toEqual({ width: "23%" });
+    });
+
+    it("should call the onClick handler when the progress bar is clicked on", () => {
```
`triggers an event when a clickable progress bar is clicked`

diego-antonelli commented on packages/pluggableWidgets/progress-bar-web/src/components/__tests__/ProgressBar.spec.tsx:
```diff
@@ -0,0 +1,215 @@
+        expect(onClickSpy).toHaveBeenCalledTimes(1);
+    });
+
+    it("should properly handle a different range", () => {
```
`handles a different range`

diego-antonelli commented on packages/pluggableWidgets/progress-bar-web/src/components/__tests__/ProgressBar.spec.tsx:
```diff
@@ -0,0 +1,215 @@
+        );
+
+        // 23 on a range from 20 to 30 is 30% progress.
+        expect(getProgressBarStyle(progressBar)).toEqual({ width: "30%" });
```
good one! 👍

diego-antonelli commented on packages/pluggableWidgets/progress-bar-web/src/components/__tests__/ProgressBar.spec.tsx:
```diff
@@ -0,0 +1,215 @@
+        expect(getProgressBarStyle(progressBar)).toEqual({ width: "30%" });
+    });
+
+    it("should clamp a current value lower than the minimum value to 0% progress", () => {
```
`clamps a current...` or even `fallbacks the current values...`

diego-antonelli commented on packages/pluggableWidgets/progress-bar-web/src/components/__tests__/ProgressBar.spec.tsx:
```diff
@@ -0,0 +1,215 @@
+        expect(getProgressBarStyle(progressBar)).toEqual({ width: "0%" });
+    });
+
+    it("should clamp a current value higher than the maximum value to 100% progress", () => {
```
Same as previous comment

diego-antonelli commented on packages/pluggableWidgets/progress-bar-web/src/components/__tests__/ProgressBar.spec.tsx:
```diff
@@ -0,0 +1,215 @@
+        expect(getProgressBarStyle(progressBar)).toEqual({ width: "100%" });
+    });
+
+    it("should not be clickable when there is no onClick handler provided", () => {
```
`is not clickable when...`

diego-antonelli commented on packages/pluggableWidgets/progress-bar-web/src/components/__tests__/ProgressBar.spec.tsx:
```diff
@@ -0,0 +1,215 @@
+        expect(progressBar.find(".progress").hasClass("widget-progress-bar-clickable")).toBe(false);
+    });
+
+    describe("should show a runtime error Alert", () => {
```
`shows a...`

diego-antonelli commented on packages/theming/atlas/src/themesource/atlas_ui_resources/web/core/widgets/_progressbar.scss:
```diff
@@ -13,7 +13,6 @@
 
 .progress {
     overflow: hidden;
-    height: 16px;
```
Nice one for custom content!! 👏

diego-antonelli reviewed (changes requested):
🎄 
Impressive job @Keraito !! Very good first PR!

_I left few small comments, you can either consider them and implement or reply with your thoughts about._

Keraito commented on packages/pluggableWidgets/progress-bar-web/src/ProgressBar.editorPreview.tsx:
```diff
@@ -0,0 +1,59 @@
+            currentValue={currentValue}
+            minValue={minValue}
+            maxValue={maxValue}
+            onClick={
```
Whether this callback is defined or not determines whether the `progress-bar-clickable` class is applied, which results in a `cursor: pointer`. Do you think that's worth keeping in design mode?

diego-antonelli replied:
Design mode doesn't have either a cursor pointer

Keraito commented on packages/pluggableWidgets/progress-bar-web/src/ProgressBar.tsx:
```diff
@@ -0,0 +1,45 @@
+        switch (props.type) {
+            case "dynamic":
+                return {
+                    currentValue: Number(props.dynamicCurrentValue?.value ?? defaultValues.currentValue),
```
I already check for this in `ProgressBar.editorConfig.jsx` at compile time. Should I add another runtime check and show an `Alert` if they are not set? (Also for the expression case)

diego-antonelli replied:
Sorry.. It was supposed to be `should`, sometimes the values are not available then if we set defaults it will just change during runtime. Better fallback as 0

Keraito replied:
Will change, but please not that setting the current value to 0 is not the same as setting the progress to 0%, since it depends on the minimum and maximum value.

diego-antonelli replied:
In this case we need to set as 0% instead of 50

Keraito replied:
@diego-antonelli 👆  just to make sure it wasn't missed

diego-antonelli replied:
No, I checked the code and looks fine

Keraito commented on packages/pluggableWidgets/progress-bar-web/src/ProgressBar.xml:
```diff
@@ -0,0 +1,106 @@
+                    <description />
+                </property>
+            </propertyGroup>
+            <propertyGroup caption="Progress Label">
```
Could you check if what I did in bcbda92 is what you meant?

diego-antonelli replied:
Exactly!

Keraito commented:
@diego-antonelli I went through your comments and either addressed it + resolved your comments or replied with follow up.

diego-antonelli commented:
> @diego-antonelli I went through your comments and either addressed it + resolved your comments or replied with follow up.

Nice, I also replied to your questions.

**PS: Do not resolve comments, otherwise the reviewer cannot see the proposed solution and check if that was really resolved 😄

Keraito commented:
> **PS: Do not resolve comments, otherwise the reviewer cannot see the proposed solution and check if that was really resolved 😄

Sorry, old habit 😄 

Also @diego-antonelli , I noticed that in the new design specs we have a small medium and large size. They respectively make the progress bar 4px 8px and 12px high. The previous one was like 16px high. But with the new size configurations, it's extremely hard to vertically align text in it because there's only so much pixels to work with. Should we do something about this?

Keraito commented:
For context:

This is small:
![image](https://user-images.githubusercontent.com/5955441/108336428-50f26380-71d4-11eb-91e4-41a4a816d2fd.png)

This is medium:
![image](https://user-images.githubusercontent.com/5955441/108336295-2accc380-71d4-11eb-9a0d-14860955a418.png)

This is large:
![image](https://user-images.githubusercontent.com/5955441/108336495-62d40680-71d4-11eb-8957-9edcdf97a4ec.png)

cjfhodges commented:
> For context this is how they currently are with already some additional adjustments.
> 
> This is small:
> ![image](https://user-images.githubusercontent.com/5955441/108336428-50f26380-71d4-11eb-91e4-41a4a816d2fd.png)
> 
> This is medium:
> ![image](https://user-images.githubusercontent.com/5955441/108336295-2accc380-71d4-11eb-9a0d-14860955a418.png)
> 
> This is large:
> ![image](https://user-images.githubusercontent.com/5955441/108336495-62d40680-71d4-11eb-8957-9edcdf97a4ec.png)

Small -> 8px (no percentage value)
Medium -> 16px (font size 12px)
Large -> 24px (font size 14px)

Keraito commented:
@diego-antonelli Addressed the remaining feedback and the new design input. Could you check whether everything has been resolved properly? Also CI is still failing on the spaces in the images that i deleted 🙃

diego-antonelli commented:
@Keraito please rebase, I've removed the images manually from master.

Keraito commented:
@diego-antonelli doneso! If this passes CI and the last comment I pinged u for is fine, I will pass this on to testing.

Keraito commented:
@diego-antonelli can you check whether I got all the files in ed39d04? Then @leonardomendix can rebase and continue 😄

diego-antonelli commented:
> @diego-antonelli can you check whether I got all the files in [ed39d04](https://github.com/mendix/widgets-resources/commit/ed39d044b2c3c7b24943423ede8020d8ef5b722d)? Then @leonardomendix can rebase and continue 😄

Thats correct

Keraito commented:
> Small -> 8px (no percentage value) 

@cjfhodges I assumed that [the above comment](https://github.com/mendix/widgets-resources/pull/455#issuecomment-781220073) applied to all label types and not only percentage, can you confirm or deny?

cjfhodges commented:
> > Small -> 8px (no percentage value)
> 
> @cjfhodges I assumed that [the above comment](https://github.com/mendix/widgets-resources/pull/455#issuecomment-781220073) applied to all label types and not only percentage, can you confirm or deny?

Yes to all types.

Keraito commented:
Discussed in standup that we will introduce a tooltip for the small version.

Keraito commented:
Update 2:
- The above is referring to built in HTML `title` attribute.
- We will only do this for textual labels.
- If custom widget label we don't show.
- Show description at the custom label attribute/property that it doesn't matter if size = small

diego-antonelli commented:
@Keraito it needs a rebase on master

Keraito commented:
Will do

Keraito commented:
@diego-antonelli could you review the following commits? Especially the first 2, and small check whether the third is fine

- e6c9688 : Added the tooltip for text and percentage label.
- 0dcb790: Had to use the inline css classname check to also hide the custom label when size is small
- 626123: now the progressbar would determine the vertical distance between itself and a potential alert. I moved it towards the alert because it's only necessary when there's an actual alert.
- 31ed7a9: Add description to the custom label property that indicates the small size behaviour

Keraito commented on packages/theming/atlas/src/themesource/atlas_ui_resources/web/core/helpers/_progressbar.scss:
```diff
@@ -33,6 +33,11 @@
     font-size: 0px;
```
Already had font-size: 0px for textual labels, so only had to add the logic for hiding custom labels below.

Keraito commented:
@diego-antonelli

- [ ] Updated the above comment to reflect the new commits. 
- [ ] Also implemented the CSS fix we discussed in 0e32e65.
- [ ] Also rebased.

Keraito commented:
This is all green again, so @leonardomendix can do this magic again 😄  Your feedback was addressed and some design updates were applied according to https://github.com/mendix/widgets-resources/pull/455#issuecomment-785813440.
